### PR TITLE
gen-host-js: instance ESM output by default

### PR DESCRIPTION
**Pull request by guybedford**
Outputs a direct ES module from the JS host generator by default that can be directly imported in JS environments.

For the custom instantiation API that was the previous output, a new `--instantiation` flag is provided.

Instance-level customization flags are also added including:
* `--compat`: Enables all compatibility mode flags (right now just `--nodejs-compat`, but based on the assumption that a new --tla-compat flag will be added in a follow-up PR).
* `--nodejs-compat`: Handle Node.js instantiation when no fetch global is present

The above flags are only compatible with instance output and will throw otherwise.

This PR also parallelizes the core Wasm instantiations, and converts the instantiation API from `instantiateCore(path, imports)` to `compileCore(path)` to enable this parallelization.

In addition a new `--map` is added with support for custom import specifier mappings that can apply to both output modes. For example - `--map "imports=./imports.js"` allows specifying that the imports should resolve to a relative JS specifier.

For testing, the first line comment of the JS file starting with `// Flags:` is used to pass these output options flags, with most tests now including `// Flags: --instantiation` for backwards-compat.

### CODE REVIEW

File: crates/gen-host-js/tests/runtime.rs
Comment (alexcrichton) [inline crates/gen-host-js/tests/runtime.rs:101]: This can perhaps be simplified with:

```rust
src.lines().next().and_then(|s| s.strip_prefix("// Flags: ")).unwrap_or("")
```

File: tests/runtime/flavorful/host.ts
Comment (alexcrichton) [inline tests/runtime/flavorful/host.ts:74]: What is "TLA"? (curious, haven't seen this before)

File: tests/runtime/flavorful/host.ts
Comment (alexcrichton) [inline tests/runtime/flavorful/host.ts:46]: Is this possible to lift to a top-level `import`?

Additionally, how come this needs `@ts-ignore`?

File: crates/gen-host-js/src/lib.rs
Comment (alexcrichton) [inline crates/gen-host-js/src/lib.rs:262]: Note to handle this before merging.

File: crates/gen-host-js/src/lib.rs
Comment (alexcrichton) [inline crates/gen-host-js/src/lib.rs:336]: Note to handle this before merging

File: crates/gen-host-js/src/lib.rs
Comment (alexcrichton) [inline crates/gen-host-js/src/lib.rs:90]: I think that you can use `#[clap(value_parser = maps_str_to_map)]` on the field in `Opts` to avoid needing to manually do the parse step here if you'd like.

File: crates/gen-host-js/src/lib.rs
Comment (alexcrichton) [inline crates/gen-host-js/src/lib.rs:800]: I think this string literal may need one tab of indentation to match.

Additionally, is it worth trying to use `instantiateStreaming` instead of just `compileStreaming` since I think at least historically that's what browsers recommended? The `LoadWasm` intrinsic here could perhaps be replaced with `InstantiateWasm` where the imports are passed in.

File: crates/gen-host-js/src/lib.rs
Comment (alexcrichton) [inline crates/gen-host-js/src/lib.rs:969]: Personally I would ideally prefer to have sort of a mish-mash sprinkling of `if self.gen.opts.instantiation` throughout since I feel that makes code difficult to read and change in the future. Would it be possible to instead change the return value of this function to something that works in both `--instantiation` and not modes? For example the `exports` here could return a form of `HashMap<???, String>` or something and then the caller would, once, check `instantiation` and create an export object or create a bunch of exported functions.

**Review by alexcrichton — COMMENTED** [review]
Looks reasonable to me!

Could you expand on the longer-term story you see fo the `--instantiation`/etc flags? My experience in `wasm-bindgen` is that the flags for "what js environment are you targeting" can get out of hand quite quickly and become unwieldy to manage internally. The solution settled on over there was an all-encompassing `--target $foo` flag where there was no other need to have lots of other various flags.

Do you expect that the set of flags here will continue to grow? Are there lots of other targets that would grow custom flags that could have odd interactions with one another? Or perhaps would a `--target` flag make sense here as well?

File: tests/runtime/flavorful/host.ts
Comment (guybedford) [reply tests/runtime/flavorful/host.ts:74]: Top-level await. Sorry forget it's not widely known!

File: tests/runtime/flavorful/host.ts
Comment (guybedford) [reply tests/runtime/flavorful/host.ts:46]: This can't be lifted as it would create a top-level await execution cycle since the module imports the JS which imports this module (to fulfill the imports). We could just create another module, but it's quite nice to encapsulate the test in a single file as well.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:411]: Well spotted. In theory you avoid one event loop tick by awaiting the promise in the async function instead of returning it. But it's completely negligible from any user perspective so that writing the shortest code is still the preference for me with these. Changing to be a non-async function for now.

Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:411]: Actually the `import('node:fs')` does need the await, so I've left it as-is for now.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:800]: Added, although as far as I can tell from the source formatter, it will remove that indentation anyway?

In theory if we wanted to instantiate the same core module multiple times, we would want the compiled but uninstantiated module. For this reason I quite like treating that as the go-to representation, and since it matches what will happen with import reflection as well.

It was just the sync former APIs which are not - that is `WebAssembly.instantiate` is not recommended when passing a buffer. But that it's overloaded form that takes a Module instance is fine to use despite the MDN warning, since the steps involved are equivalent from a spec perspective at least.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:842]: The pattern I'm using for all intrinsics and top-level definitions since the formatting PR to ensure consistent double line spacing is to have a newline both before and after when they should be double spaced between eachother (and that's why there's the strip leading newline behaviour in the final write). Otherwise double spacing after each item leads to double spacing between items that intend to be single-spaced. I'm open to better layout suggestions though for sure! Updated to use `\n` for now.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:833]: Yes, there's no restriction they would be the first statement types. That said if we did have that restriction module parsing would have been a lot easier...!

Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:833]: I do plan to add grouping to this output like we did with the C generator, just wanted to get the functionality right first.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:2388]: Thanks.

**Comment by guybedford** [comment]
In terms of the compat / target flags, one thing I don't like about versioning targets is it puts too much responsibility on the user to know their exact platform support, whereas having good defaults and some hints / options can get you 99% of the way.

I guess the model I'm thinking of is that the default output works in most environments, then by adding `--compat` you get something that works everywhere, or by digging into the specific `--x-compat` flags you can work out your own kind of compat matrix as necessary.

The features we're branching on here are currently really just:
* Top-level await support (~90% of browsers)
* fetch support (basically everywhere except Node.js)
* Node.js support (using readfile as a replacement for fetch)

That's really it as far as the current status is.

In future we might have import reflection to import the core modules instead of compileStreaming, but apart from that I think most other things would be at the level of core Wasm itself. Eg if we want to lean more heavily on new core Wasm features while still emitting legacy core modules and branching on that. But in theory even something like that could be captured by the current `--compat` system?

I'm not sure though, I may be going down that same path of progressive complexity here too, so would be interested your thoughts further on this and what you would have in mind for the above in a target-based system.

File: tests/runtime/flavorful/host.ts
Comment (guybedford) [reply tests/runtime/flavorful/host.ts:46]: Turns out the `@ts-ignore` wasn't necessary.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs]: For now I've just exported the imports typings so they are at least available.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs:964]: By switching to namespaces there is now much more consistent handling between the modes. Overall I think this gating is a semantically important one as it represents the difference between exporting a namespace and a top-level declaration.

**Review by alexcrichton — APPROVED** [review]
Ok sounds reasonable to me, I'm happy to go with these flags for now and see how it ends up.

As part of this PR or as a follow up, would you be up for plumbing the new JS options into checkboxes and such for the demo page? The Wasmtime output has an example of this and I think it might be neat to get a demo online of the different outputs.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [reply crates/gen-host-js/src/lib.rs]: I've added this hoisting now.

File: crates/gen-host-js/src/lib.rs
Comment (guybedford) [inline crates/gen-host-js/src/lib.rs:743]: I've added a new parallel instantiation here, and also switched `instantiateCore` to `compileCore` so that it can be called in parallel (since import objects otherwise require ordered instantiation). Please let me know if this model works for you?

**Comment by guybedford** [comment]
I've added the demo UI options for "Compat" and "Custom Instantiation". Will merge when CI passes.

**Pull request merged**
